[WIP] Introduce RPC subsystem.

AndreiPashkin opened the pull request:
# OneDrive Client - RPC

The project's RPC system is defined in this place.

## Motivation

This sub-project is created as a consequence of dissatisfaction with ZeroMQ and other RPC systems. The purpose of it is to build a reliable RPC system, which would handle peers disconnects reliably and provide message-delivery guarantees.
Another reason is desire to have freedom in choosing the technology-stack.

## Features

-   Reliability. Particularly - reliable handling of peers disconnects
    using heartbeats.
-   High-level interface for common communication patterns:
    -   Request/Response.
    -   Publish/Subscribe.
    -   Producer/Consumer.
    -   Broadcast.
    -   Notification.

-   Basic validation of messages.
-   Pluggable transports.
-   Language-independent contracts.
-   Automatic generation of clients and servers for the services.

## Architecture

The architecture consists of three layers: 
- Transport layer. 
- RPC layer. 
- User-defined client/server layer.

### Transport layer

Responsible for consistent and guaranteed delivery of messages as byte sequences.

Transport layer abstracts concrete method of message-passing which could be TCP, UDP, ZeroMQ, D-Bus, HTTP, etc.

### RPC layer

Responsible for: 
- Maintaining connection with peers: 
    - Tracking whether peers are "alive" or "dead" using heartbeats. 
    - Tracking acknowledgements of messages when it's required. 
- Implementing common communication patterns and providing high-level interface for using them.

RPC Layer is dependent on transport layer. At the same it should not depend on any implementation details of particular transport implementation.

### User-defined client/server layer.

Responsible for providing stubs for servers and client for concrete services of the application. Both are supposed to be thin wrappers around RPC layer that route calls of concrete methods ("get\_item()", "put()", etc) to correct RPC calls made by methods implemented in RPC layer.

Generated clients are fully-functional accessors for associated services. Server-stubs are server implementations that only lack business logic and otherwise are fully-functional servers that can accept connections. Business logic for server-stubs are supposed to be provided via call-backs by users of the stub-classes.

Client and server-stub generation is supposed to be made based on language-independent interface declarations (or
contracts).

## Technology stack

TCP - primary transport protocol. Protobuf - as an interface declaration language, a tool for code-generation and serialization.

Commit:
RPC: start RPC-subproject.

Commit:
RPC: setup Python package.

Commit:
RPC: introduce Transport base class.

Commit:
RPC: introduce TCP-based Transport implementation.

iivanou commented on .travis.yml:
```diff
@@ -24,3 +24,9 @@ matrix:
+        install:
+          - pip install 'tox>=2.4.1,<3'
+        script:
+          - cd rpc/pyton && tox
```
Probably "python" instead of "pyton"?

iivanou commented on rpc/python/src/onedrive_client/rpc/transport/tcp.py:
```diff
@@ -0,0 +1,299 @@
+        )
+        self.__closed = False
+
+    @property
```
Private property?

iivanou commented on rpc/python/src/onedrive_client/rpc/transport/tcp.py:
```diff
@@ -0,0 +1,299 @@
+
+            while True:
+                message = self.__extract_message(buffer)
+                if message is not None:
```
if message:

iivanou commented on rpc/python/tests/utils/gevent.py:
```diff
@@ -0,0 +1,16 @@
+import gevent
```
Doc strings are missing

iivanou reviewed (changes requested):
Overall looks good. I would add an example how to use it into the comments.

AndreiPashkin commented:
@Almost1990 
> Overall looks good. I would add an example how to use it into the comments.

Sure.

Commit:
RPC: disable "bad-continuation" Pylint check.

Because it thinks that this is badly formatted:

    if (condition() or
        timeout is not None and
        (time.time() - timestamp) > timeout):
        return

And that the second line must be indented.

Commit:
RPC: exempt functions shorter than 4 lines from having a docstring.

Commit:
RPC: make Pylint allow classes with more than 7 attributes.

The value is increased up to 15.

Commit:
RPC: Make PEP8 E129 be ignored.

Because it treats this formatting as invalid:

    if (condition() or
        timeout is not None and
        (time.time() - timestamp) > timeout):
        return

It says that the second line is underindented.

Commit:
RPC: introduce boolean expressions interpreter.

Commit:
RPC: increase maximum allowed return statements in Pylint config.

Commit:
RPC: loosen restrictions on module names in Pylint config.

Commit:
RPC: loosen methods name restrictions in Pylint config.

Commit:
RPC: loosen restrictions on arguments number in Pylint config.

Commit:
entities: remove excessive namespace nesting in Protoc-plugin name.

Commit:
RPC: progress
